### PR TITLE
Remove S3 output stream

### PR DESCRIPTION
**Pull request by tlrx**
The `S3OutputStream` class has been added in elasticsearch 1.4 in order to take advantage of the AWS Multipart Upload API. At that time, the Snapshot/Restore API didn't communicate the size of the blobs to be written to the repository implementation: the logic was to open an `Outputstream` and start to write bytes to it. With this logic, we decided to bufferize the bytes in memory until we know which API to use between a single upload request or multipart upload requests. This is why the buffer and associated logic was implemented as an ` OutputStream`.

Now the blob size information is available before writing anything, the repository implementation can know upfront what will be the more suitable API to upload the blob to S3.

This pull request removes the `DefaultS3OutputStream` and `S3OutputStream` classes and moves the implementation of the upload logic directly in the `S3BlobContainer`. I think it's easier to understand and easier to maintain. It also avoids the internal buffering by passing the `InputStream` directly to the S3 client that takes care of buffering (up to 16Mb) and retry the requests. It removes pressure on memory (some allocations outside TLAB drop from 2.7Gb to 57Mb in my tests while total snapshot time drops from 35%), specially on nodes with small heap like 1Gb-4Gb.

Note: this has been tested together with #27278.

### CODE REVIEW

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java [hunk 1]
```diff
@@ -91,9 +106,12 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize) t
         if (blobExists(blobName)) {
             throw new FileAlreadyExistsException("blob [" + blobName + "] already exists, cannot overwrite");
         }
-        try (OutputStream stream = createOutput(blobName)) {
-            Streams.copy(inputStream, stream);
-        }
+
+        final UploadMethod method = (blobSize <= blobStore.bufferSizeInBytes()) ? this::executeSingleUpload : this::executeMultipartUpload;
```
Comment (imotov) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java]: I fell like a functional interface here doesn't really buy us anything comparing to a simple if statement with two calls.

**Review by imotov — APPROVED** [review]
LGTM from the snapshot/restore perspective. Love the simplification. Left one minor observation. Feel free to ignore it if you disagree.

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java (hunk 1 above)
Comment (tlrx) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java]: I agree and pushed https://github.com/elastic/elasticsearch/pull/27280/commits/4421ee1e0c876307a6f5da795894533d05d60859

**Comment by tlrx** [comment]
@elasticmachine retest this please

**Review by ywelsch — APPROVED** [review]
LGTM. Left one minor nit.

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 2]
```diff
@@ -87,7 +87,7 @@
      * use of the Multipart API and may result in upload errors. Defaults to the minimum between 100MB and 5% of the heap size.
      */
     static final Setting<ByteSizeValue> BUFFER_SIZE_SETTING = Setting.byteSizeSetting("buffer_size", DEFAULT_BUFFER_SIZE,
-            new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.TB));
+            new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.GB));
```
[the remark this answers is not in the document]
Comment (tlrx) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: I pushed https://github.com/elastic/elasticsearch/pull/27280/commits/4aae0ceb794233f9a0835b3596865cb983bf3c79

[the remark this answers is not in the document]
Comment (ywelsch) [reply plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: Looks good. Thanks :)

**Comment by tlrx** [comment]
@elasticmachine retest this please

**Comment by tlrx** [comment]
This change has been backported to 5.6.5, 6.0.1, 6.1.0 and master.

**Pull request merged**